src/Region.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * LSST Data Management System
+ * Copyright 2014-2015 AURA/LSST.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_SG_REGION_H_
+#define LSST_SG_REGION_H_
+
+/// \file
+/// \author Serge Monkewitz
+/// \brief This file defines an interface for spherical regions.
+
+namespace lsst {
+namespace sg {
+
+// Forward declarations
+class Box;
+class Circle;
+class ConvexPolygon;
+class Ellipse;
+class UnitVector3d;
+
+/// `Region` is a minimal interface for 2-dimensional regions on the unit
+/// sphere. It provides two core pieces of functionality:
+///
+/// - It allows a region to be approximated with a simpler one.
+/// - It allows for inexact computation of the spatial relationships
+///   between two regions.
+///
+/// Given a partitioning of the unit sphere with partitions that can be
+/// bounded by Regions, this provides all the necessary functionality for
+/// determining which partitions may intersect a Region.
+///
+/// When implementing a new concrete region subclass R, the Region interface
+/// should be extended with:
+///
+///     virtual int relate(R const &) const = 0;
+///
+/// All other other Region subclasses must then implement this method. In
+/// addition, R is expected to contain the following implementation of the
+/// generic relate method:
+///
+///     virtual int relate(Region const & r) const {
+///         return invertSpatialRelations(r.relate(*this));
+///     }
+///
+/// where invertSpatialRelations is defined in SpatialRelations.h.
+///
+/// Given two Region references r1 and r2 of type R1 and R2, the net effect
+/// is that r1.relate(r2) will end up calling R2::relate(R1 const &). In other
+/// words, the call is polymorphic in the types of both r1 and r2.
+///
+/// One negative consequence of this design is that one cannot implement new
+/// Region types outside of this library.
+class Region {
+public:
+    virtual ~Region() {}
+    /// `clone` returns a deep copy of this region.
+    virtual Region * clone() const = 0;
```

Comment:
I think the proscription against C++11 in headers is irrelevant given that the motiviation behind that is just to ensure we don't confuse Swig (and there's no Swig here).  Like so many other issues (I'll try to put together a list at the end of the revue), you should get sign off on this from K-T, but I'd say you should just use `unique_ptr` here.

IMO, `shared_ptr(r->clone())` isn't so terrible on its own, because there's no way anyone would write a `clone()` method that returns an internal reference, but as soon as you have any other methods that return raw pointers, it's just too easy to assume the wrong thing.